File: python/astroengine/vedic.py
"""astroengine.vedic -- the Vedic/Jyotish layer: nakshatras and the Vimshottari
dasha, built on the already-validated sidereal longitudes.

A nakshatra is one of 27 equal lunar mansions of 13 deg 20' (360/27) in the
sidereal zodiac, each divided into four padas of 3 deg 20'. Each nakshatra has a
ruling planet, cycling Ketu, Venus, Sun, Moon, Mars, Rahu, Jupiter, Saturn,
Mercury (nine lords, three times around). The Vimshottari dasha is a 120-year
sequence of planetary periods (mahadashas) in that same order, with fixed
lengths summing to 120; the starting dasha is the lord of the Moon's birth
nakshatra, and the portion already elapsed is the fraction of the nakshatra the
Moon has traversed. Each mahadasha subdivides into antardashas (and those into
pratyantardashas) of the nine lords, length proportional to the lord's years.

Nakshatra placement is exact division of the sidereal longitude the engine
already validates against Swiss Ephemeris. The dasha is deterministic time
arithmetic; the period year is a fixed length (365.25 days, the common Jyotish
convention) by default. The TS port (vedic.ts) reproduces every value and the
golden fixtures pin the two together.

Floor division uses ``math.floor(x / span)`` rather than the ``//`` operator so
it matches JavaScript's ``Math.floor`` bit for bit on the exact sign/nakshatra
boundaries (Python ``//`` applies a correction that can land a boundary value in
the lower bucket).
"""
import math

NAKSHATRAS = [
    "Ashwini", "Bharani", "Krittika", "Rohini", "Mrigashira", "Ardra",
    "Punarvasu", "Pushya", "Ashlesha", "Magha", "Purva Phalguni",
    "Uttara Phalguni", "Hasta", "Chitra", "Swati", "Vishakha", "Anuradha",
    "Jyeshtha", "Mula", "Purva Ashadha", "Uttara Ashadha", "Shravana",
    "Dhanishta", "Shatabhisha", "Purva Bhadrapada", "Uttara Bhadrapada", "Revati",
]

# The Vimshottari order and each lord's period in years (totalling 120).
VIMSHOTTARI_ORDER = ["ketu", "venus", "sun", "moon", "mars", "rahu",
                     "jupiter", "saturn", "mercury"]
VIMSHOTTARI_YEARS = {"ketu": 7, "venus": 20, "sun": 6, "moon": 10, "mars": 7,
                     "rahu": 18, "jupiter": 16, "saturn": 19, "mercury": 17}

NAK_SPAN = 360.0 / 27.0          # 13 deg 20'
VIMSHOTTARI_TOTAL = 120.0
DASHA_YEAR = 365.25              # days per dasha-year (common Jyotish convention)


def nakshatra(sidereal_lon):
    """The nakshatra of a sidereal longitude: index (0 = Ashwini), name, pada
    (1..4), ruling (dasha) lord, and degrees into the nakshatra."""
    lon = sidereal_lon % 360.0
    i = math.floor(lon / NAK_SPAN) % 27
    pos = lon - i * NAK_SPAN
    pada = math.floor(pos / (NAK_SPAN / 4.0)) + 1
    return {"index": i, "name": NAKSHATRAS[i], "pada": pada,
            "lord": VIMSHOTTARI_ORDER[i % 9], "pos": pos}


def nakshatra_at(engine, jd_ut, body="moon", zodiac="sidereal:lahiri"):
    """The nakshatra of a body (default the Moon) at jd, in a sidereal zodiac."""
    return nakshatra(engine.longitude(body, jd_ut, zodiac=zodiac))


def vimshottari_dashas(moon_lon, natal_jd, levels=2, year_length=DASHA_YEAR,
                       count=9):
    """The Vimshottari dasha timeline from the Moon's sidereal longitude. Returns
    {start_lord, balance_years, dashas}, where dashas is `count` mahadashas
    (the first beginning before birth, with its elapsed portion), each with
    antardashas when levels >= 2. Periods are in true Julian Days (UT)."""
    nak_i = math.floor((moon_lon % 360.0) / NAK_SPAN) % 27
    pos = (moon_lon % 360.0) - nak_i * NAK_SPAN
    start_lord = VIMSHOTTARI_ORDER[nak_i % 9]
    elapsed = pos / NAK_SPAN
    y0 = VIMSHOTTARI_YEARS[start_lord]
    li = VIMSHOTTARI_ORDER.index(start_lord)
    t = natal_jd - elapsed * y0 * year_length     # the first mahadasha's true start
    dashas = []
    for k in range(count):
        lord = VIMSHOTTARI_ORDER[(li + k) % 9]
        years = VIMSHOTTARI_YEARS[lord]
        span = years * year_length
        maha = {"level": 1, "lord": lord, "start": t, "end": t + span, "sub": []}
        if levels >= 2:
            sli = VIMSHOTTARI_ORDER.index(lord)
            st = t
            for j in range(9):
                sl = VIMSHOTTARI_ORDER[(sli + j) % 9]
                sub_span = (years * VIMSHOTTARI_YEARS[sl] / VIMSHOTTARI_TOTAL) * year_length
                maha["sub"].append({"lord": sl, "start": st, "end": st + sub_span})
                st += sub_span
        dashas.append(maha)
        t += span
    return {"start_lord": start_lord, "balance_years": (1.0 - elapsed) * y0,
            "dashas": dashas}


def _active_in(periods, target):
    """The period in a tiled list containing target, or None."""
    for p in periods:
        if p["start"] <= target < p["end"]:
            return p
    return None


def vimshottari_active(moon_lon, natal_jd, target_jd, year_length=DASHA_YEAR):
    """The mahadasha, antardasha, and pratyantardasha lords active at target_jd.
    None before the first mahadasha begins."""
    # enough mahadashas to cover any reasonable target (one full 120y cycle + 1)
    timeline = vimshottari_dashas(moon_lon, natal_jd, levels=2,
                                  year_length=year_length, count=10)["dashas"]
    maha = _active_in(timeline, target_jd)
    if maha is None:
        return None
    antar = _active_in(maha["sub"], target_jd)
    if antar is None:
        return {"maha": maha["lord"], "antar": None, "pratyantar": None}
    # pratyantardashas tile the antardasha among the nine lords from its lord
    ay = VIMSHOTTARI_YEARS[maha["lord"]] * VIMSHOTTARI_YEARS[antar["lord"]] / VIMSHOTTARI_TOTAL
    sli = VIMSHOTTARI_ORDER.index(antar["lord"])
    st = antar["start"]
    pratyantar = None
    for j in range(9):
        sl = VIMSHOTTARI_ORDER[(sli + j) % 9]
        span = (ay * VIMSHOTTARI_YEARS[sl] / VIMSHOTTARI_TOTAL) * year_length
        if st <= target_jd < st + span:
            pratyantar = sl
            break
        st += span
    return {"maha": maha["lord"], "antar": antar["lord"], "pratyantar": pratyantar}


def vimshottari_at(engine, natal_jd, target_jd, zodiac="sidereal:lahiri",
                   year_length=DASHA_YEAR):
    """Vimshottari dasha active at target_jd, from the natal Moon's nakshatra.
    Returns {moon_nakshatra, moon_pada, start_lord, maha, antar, pratyantar}."""
    moon_lon = engine.longitude("moon", natal_jd, zodiac=zodiac)
    nak = nakshatra(moon_lon)
    active = vimshottari_active(moon_lon, natal_jd, target_jd, year_length) or {}
    return {"moon_nakshatra": nak["name"], "moon_pada": nak["pada"],
            "start_lord": VIMSHOTTARI_ORDER[nak["index"] % 9], **active}

File: packages/caelus/src/vedic.ts
/**
 * astroengine vedic -- the Vedic/Jyotish layer: nakshatras and the Vimshottari
 * dasha, built on the already-validated sidereal longitudes.
 *
 * A nakshatra is one of 27 equal lunar mansions of 13 deg 20' in the sidereal
 * zodiac, each with four padas and a ruling planet cycling Ketu, Venus, Sun,
 * Moon, Mars, Rahu, Jupiter, Saturn, Mercury. The Vimshottari dasha is a
 * 120-year sequence of planetary periods in that order; the starting dasha is
 * the lord of the Moon's birth nakshatra, with the elapsed portion set by how
 * far the Moon has moved through it. Mahadashas subdivide into antardashas and
 * pratyantardashas of the nine lords, proportional to their years. Nakshatra
 * placement is exact division of the sidereal longitude; the dasha year is a
 * fixed 365.25 days by default (the common Jyotish convention). Mirrors the
 * Python reference (astroengine/vedic.py); the golden fixtures pin the two.
 */
import { Engine, BodyId, Zodiac } from "./chart.js";

export const NAKSHATRAS = [
  "Ashwini", "Bharani", "Krittika", "Rohini", "Mrigashira", "Ardra",
  "Punarvasu", "Pushya", "Ashlesha", "Magha", "Purva Phalguni",
  "Uttara Phalguni", "Hasta", "Chitra", "Swati", "Vishakha", "Anuradha",
  "Jyeshtha", "Mula", "Purva Ashadha", "Uttara Ashadha", "Shravana",
  "Dhanishta", "Shatabhisha", "Purva Bhadrapada", "Uttara Bhadrapada", "Revati",
] as const;

/** The Vimshottari order and each lord's period in years (totalling 120). */
export const VIMSHOTTARI_ORDER = [
  "ketu", "venus", "sun", "moon", "mars", "rahu", "jupiter", "saturn", "mercury",
] as const;
export const VIMSHOTTARI_YEARS: Record<(typeof VIMSHOTTARI_ORDER)[number], number> = {
  ketu: 7, venus: 20, sun: 6, moon: 10, mars: 7, rahu: 18, jupiter: 16, saturn: 19, mercury: 17,
};

export const NAK_SPAN = 360 / 27; // 13 deg 20'
const VIMSHOTTARI_TOTAL = 120;
export const DASHA_YEAR = 365.25; // days per dasha-year (common Jyotish convention)

function mod360(x: number): number {
  return ((x % 360) + 360) % 360;
}

export interface Nakshatra {
  index: number;
  name: string;
  pada: number;
  lord: string;
  /** Degrees into the nakshatra, 0..13.333. */
  pos: number;
}

/** The nakshatra of a sidereal longitude. */
export function nakshatra(siderealLon: number): Nakshatra {
  const lon = mod360(siderealLon);
  const i = Math.floor(lon / NAK_SPAN) % 27;
  const pos = lon - i * NAK_SPAN;
  const pada = Math.floor(pos / (NAK_SPAN / 4)) + 1;
  return { index: i, name: NAKSHATRAS[i], pada, lord: VIMSHOTTARI_ORDER[i % 9], pos };
}

/** The nakshatra of a body (default the Moon) at jd, in a sidereal zodiac. */
export function nakshatraAt(
  engine: Engine, jdUt: number, body: BodyId = "moon", zodiac: Zodiac = "sidereal:lahiri",
): Nakshatra {
  return nakshatra(engine.longitude(body, jdUt, { zodiac }));
}

export interface DashaSub { lord: string; start: number; end: number; }
export interface Dasha { level: number; lord: string; start: number; end: number; sub: DashaSub[]; }
export interface DashaTimeline { start_lord: string; balance_years: number; dashas: Dasha[]; }

/** The Vimshottari dasha timeline from the Moon's sidereal longitude. */
export function vimshottariDashas(
  moonLon: number, natalJd: number, levels = 2, yearLength = DASHA_YEAR, count = 9,
): DashaTimeline {
  const lon = mod360(moonLon);
  const nakI = Math.floor(lon / NAK_SPAN) % 27;
  const pos = lon - nakI * NAK_SPAN;
  const startLord = VIMSHOTTARI_ORDER[nakI % 9];
  const elapsed = pos / NAK_SPAN;
  const y0 = VIMSHOTTARI_YEARS[startLord];
  const li = (VIMSHOTTARI_ORDER as readonly string[]).indexOf(startLord);
  let t = natalJd - elapsed * y0 * yearLength;
  const dashas: Dasha[] = [];
  for (let k = 0; k < count; k++) {
    const lord = VIMSHOTTARI_ORDER[(li + k) % 9];
    const years = VIMSHOTTARI_YEARS[lord];
    const span = years * yearLength;
    const maha: Dasha = { level: 1, lord, start: t, end: t + span, sub: [] };
    if (levels >= 2) {
      const sli = (VIMSHOTTARI_ORDER as readonly string[]).indexOf(lord);
      let st = t;
      for (let j = 0; j < 9; j++) {
        const sl = VIMSHOTTARI_ORDER[(sli + j) % 9];
        const subSpan = (years * VIMSHOTTARI_YEARS[sl] / VIMSHOTTARI_TOTAL) * yearLength;
        maha.sub.push({ lord: sl, start: st, end: st + subSpan });
        st += subSpan;
      }
    }
    dashas.push(maha);
    t += span;
  }
  return { start_lord: startLord, balance_years: (1 - elapsed) * y0, dashas };
}

function activeIn<T extends { start: number; end: number }>(periods: T[], target: number): T | null {
  return periods.find((p) => p.start <= target && target < p.end) ?? null;
}

export interface DashaActive { maha: string; antar: string | null; pratyantar: string | null; }

/** The mahadasha, antardasha, and pratyantardasha lords active at targetJd. */
export function vimshottariActive(
  moonLon: number, natalJd: number, targetJd: number, yearLength = DASHA_YEAR,
): DashaActive | null {
  const timeline = vimshottariDashas(moonLon, natalJd, 2, yearLength, 10).dashas;
  const maha = activeIn(timeline, targetJd);
  if (maha === null) return null;
  const antar = activeIn(maha.sub, targetJd);
  if (antar === null) return { maha: maha.lord, antar: null, pratyantar: null };
  const ay = VIMSHOTTARI_YEARS[maha.lord as keyof typeof VIMSHOTTARI_YEARS]
    * VIMSHOTTARI_YEARS[antar.lord as keyof typeof VIMSHOTTARI_YEARS] / VIMSHOTTARI_TOTAL;
  const sli = (VIMSHOTTARI_ORDER as readonly string[]).indexOf(antar.lord);
  let st = antar.start;
  let pratyantar: string | null = null;
  for (let j = 0; j < 9; j++) {
    const sl = VIMSHOTTARI_ORDER[(sli + j) % 9];
    const span = (ay * VIMSHOTTARI_YEARS[sl] / VIMSHOTTARI_TOTAL) * yearLength;
    if (st <= targetJd && targetJd < st + span) { pratyantar = sl; break; }
    st += span;
  }
  return { maha: maha.lord, antar: antar.lord, pratyantar };
}

/** Vimshottari dasha active at targetJd, from the natal Moon's nakshatra. */
export function vimshottariAt(
  engine: Engine, natalJd: number, targetJd: number,
  zodiac: Zodiac = "sidereal:lahiri", yearLength = DASHA_YEAR,
): { moon_nakshatra: string; moon_pada: number; start_lord: string } & Partial<DashaActive> {
  const moonLon = engine.longitude("moon", natalJd, { zodiac });
  const nak = nakshatra(moonLon);
  const active = vimshottariActive(moonLon, natalJd, targetJd, yearLength) ?? {};
  return {
    moon_nakshatra: nak.name, moon_pada: nak.pada,
    start_lord: VIMSHOTTARI_ORDER[nak.index % 9], ...active,
  };
}
